src/core/document.js

```diff
@@ -1241,29 +1243,35 @@ class PDFDocument {
       return shadow(this, "fieldObjects", Promise.resolve(null));
     }
 
-    const allFields = Object.create(null);
-    const fieldPromises = new Map();
-    for (const fieldRef of this.catalog.acroForm.get("Fields")) {
-      this._collectFieldObjects("", fieldRef, fieldPromises);
-    }
+    const fieldObjectsPromise = this.pdfManager
+      .ensureCatalog("toplevelPagesDict")
+      .then(pages => {
+        let kids = pages.getArray("Kids");
+        kids = Array.isArray(kids) ? kids : [];
+        const pageRefs = new Map(
+          kids.map((x, i) => [isRef(x) ? x.toString() : "", i])
+        );
```

Comment:
> I was thinking to something similar but with a cache to avoid too much calls to `getPageIndex`

How much of an issue is this going to be in real-world documents, I wonder? Obviously we can add a cache, possibly in `Catalog.getPageIndex`, but I'm not sure how necessary it actually is right now.

 > finally I was wondering if it'd be better to just make one loop to generate a map pageRef => pageIndex which could be useful in `getPageIndex` itself.

Personally, I'd strongly suggest that we do *NOT* do that!
There's no guarantee that every page of a document has form elements, and generally with all the lazy initialization/parsing that we do it seems *wrong* to force the *entire* /Pages-tree to be loading/parsed upfront like that (and it'd basically end up "breaking" to whole idea of e.g. `disableAutoFetch`).